metakb/query.py

```diff
@@ -307,27 +331,336 @@ def search(self, variation='', disease='', therapy='', gene='',
+            response['variation_descriptor'] = \
+                self._get_variation_descriptor(response, node)
+        elif label == 'TherapyDescriptor':
+            response['therapy_descriptor'] = \
+                self._get_therapy_descriptor(node)
+        elif label == 'DiseaseDescriptor':
+            response['disease_descriptor'] = self._get_disease_descriptor(node)
+        elif label == 'GeneDescriptor':
+            response['gene_descriptor'] = \
+                self._get_gene_descriptor(node, self._get_gene_value_object(node))  # noqa: E501
+        elif label == 'Document':
+            document = self._get_document(node)
+            if document:
+                response['document'] = document
+        elif label == 'Method':
+            response['method'] = self._get_method(node)
+
+        session.close()
+        return SearchIDService(**response).dict(
+            by_alias=True, exclude_none=True)
 
-    def _add_variation_descriptor(self, response, variation_descriptor,
-                                  by_id=False):
-        """Add variation descriptor to response.
+    def search_statements(
+            self, variation: str = '', disease: str = '', therapy: str = '',
+            gene: str = '', statement_id: str = ''):
+        """Get nested statements from queried concepts
 
-        :param dict response: The search response
-        :param Node variation_descriptor: Variation Descriptor Node
+        :param str variation: Variation query
+        :param str disease: Disease query
+        :param str therapy: Therapy query
+        :param str gene: Gene query
+        :param str statement_id: Statement ID query
+        :return: A dictionary containing the statements with nested
+            propositions, descriptors, methods, and supported by documents
+        """
+        response: Dict = {
+            'query': {
+                'variation': None,
+                'disease': None,
+                'therapy': None,
+                'gene': None,
+                'statement_id': None
+            },
+            'warnings': [],
+            'matches': {
+                "statements": [],
+                "propositions": []
+            },
+            'statements': []
+        }
+
+        normalized_terms = self.get_normalized_terms(
+            variation, disease, therapy, gene, statement_id, response)
+        if normalized_terms is None:
+            return SearchStatementsService(**response).dict()
+        (normalized_variation, normalized_disease,
+         normalized_therapy, normalized_gene, statement,
+         valid_statement_id) = normalized_terms
+
+        session = self.driver.session()
+        statement_nodes = list()
+        proposition_nodes = session.read_transaction(
+            self._get_propositions, normalized_therapy,
+            normalized_variation, normalized_disease, normalized_gene,
+            valid_statement_id
+        )
+
+        proposition_cache = dict()
```

Comment:
It might be worth looking at abstracting away the method calls that populate each cache and wrap them with lru_cache or something similar, rather than manually building/destroying caches in each query